### PR TITLE
service/dap: refine teardown logic

### PR DESCRIPTION
**Pull request by polinasok**
Updates #1515

### CODE REVIEW

File: service/dap/server.go
Comment (polinasok) [inline service/dap/server.go:668]: Why doesn't rpc server halt in stop()?
https://github.com/go-delve/delve/blob/c223ef656da3ddbab3a58ea87108d4217731a01c/service/rpccommon/server.go#L91-L97? 
On a related note, the terminology collision here is unfortunate: https://github.com/go-delve/delve/blob/master/Documentation/api/ClientHowto.md#gracefully-ending-the-debug-session
From the RPC client point of view, it appears that
- Detach = detach debugger from the process (maybe kill it) + close client connection
- Disconnect = close client connection
But the client how-to uses a phrase "disconnecting a running program" as a case where Detach is called. And then we have DAP where disconnect request is sent to "disconnect from the debuggee and to terminate the debug adapter". Sigh.

**Comment by polinasok** [comment]
Test failure looks unrelated:

**Comment by polinasok** [comment]
cc @hyangah

File: service/dap/server.go
Comment (hyangah) [reply service/dap/server.go:676]: Wild guess - isn't it because it's the rpc client who is supposed to call Halt (or Disconnect) command  before the server stops. Not sure about the case where the debugging stops because client unexpectedly crashes or gets disconnected.

File: service/dap/server.go
Comment (hyangah) [inline service/dap/server.go:91]: fyi - Since there is no mutex annotation in go, frequently used convention is to lay out the fields protected by a mutex after the mutex. For example, once you we start protecting debugger or conn with mu, place them after mu.

File: service/dap/server.go
Comment (hyangah) [inline service/dap/server.go:187]: can we use defer (after line 172)?

File: service/dap/server.go
Comment (hyangah) [inline service/dap/server.go:671]: Will the following work? (Sorry - it's hard for me to use github codereview ui :-( )

defer s.triggerServerStop()
s.mu.Lock()
defer s.mu.Unlock()
if s.debugger != nil {
   ... 
   s.stopDebugSession(killProcess)
 } else {
   s.stopNodebugProcess()
 }

File: service/dap/server.go
Comment (hyangah) [inline service/dap/server.go:680]: s/else if/if

What other errors are possible and why is it ok not to Detach?

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go:676]: That's precisely the case I was thinking about. The client takes care of Halt before Detach. But if there is an error reading rpc messages from client, the server will go into shutdown mode and call Stop() without halting.

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go:57]: done

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go:87]: done

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go]: done

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go]: done

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go]: done

**Review by polinasok — COMMENTED** [review]
PTAL

**Comment by polinasok** [comment]
Failing tests are unrelated:
TestIssue419github.com/go-delve/delve/pkg/proc
TestAttachDetachgithub.com/go-delve/delve/pkg/proc

File: service/dap/server.go
Comment (hyangah) [inline service/dap/server.go:869]: This lock lives beyond the if block (so affect line 884-885) which isn't obvious. 
Consider either narrowing the scope of the lock, so lock is released before going out of the if block, or
handle if mode != "local" case first 

if mode != "local" {
  s.sendErrorResponse(...)
  return
}
// mode == "local"
...

**Review by hyangah — COMMENTED** [review]
Thanks!
I don't know enough to review the Halt + Detach logic so still need review from delve maintainers. Other parts look good to me.

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go:869]: Will do. I am actually not happy how much logic ends up inside of lock in general :(

File: service/dap/server.go
Comment (polinasok) [inline service/dap/server.go:614]: @hyangah Was this meant to be regular unlock to minimize scope (hence the temp var p var)? With my new change, the entire function gets locked (so I shouldn't need `p` anymore). Not ideal? But at the same time, what else is there to do when we are existing?

Comment (hyangah) [reply service/dap/server.go:614]: Heh, this is a bug - I meant s.mu.Unlock(), not defer s.mu.Unlock. :-)

Yeah, the intention was to reduce the scope - so the goroutine that's blocked onLaunchRequest (waiting for the termination of s.nodebugProcess) can proceed as soon as it's unblocked when p's killed. But you're right. We don't need to be too clever here.

File: service/dap/server.go
Comment (aarzilli) [reply service/dap/server.go:682]: I don't think there's a particular reason, halting would probably be better. And the client howto should say detach where it says ""disconnecting a running program".

**Review by aarzilli — APPROVED** [review]
LGTM

File: service/dap/server.go
Comment (polinasok) [reply service/dap/server.go:614]: Thanks for clarifying. I remove the temp var "p".

**Comment by polinasok** [comment]
@suzmue 
TestStepOutPreservesGoroutine is failing because it hit program termination when it expected to stop after a step.
This is not related to my comment and nodebug changes. I can't reproduce it locally - a race in the test? I think it would help to enable --log-output=dap,debugger in these tests, so failed logs have more info.

**Comment by polinasok** [comment]
> @suzmue
> TestStepOutPreservesGoroutine is failing because it hit program termination when it expected to stop after a step.
> This is not related to my comment and nodebug changes. I can't reproduce it locally - a race in the test? I think it would help to enable --log-output=dap,debugger in these tests, so failed logs have more info.

Fix at https://github.com/go-delve/delve/pull/2417

**Review by hyangah — APPROVED** [review]

**Comment by derekparker** [comment]
Needs rebase, I'll TAL tomorrow.

**Comment by polinasok** [comment]
> Needs rebase, I'll TAL tomorrow.

I am working on the merge.

File: service/dap/server.go
Comment (polinasok) [inline service/dap/server.go:918]: I am not sure why this did not show up before the merge, but there was a test with negative attach pid that resulted in a panic from debugger.New that skipped releasing the lock and then disconnect was forever waiting to acquire it. I changed onLaunchRequest in a similar way as well just in case.

**Comment by polinasok** [comment]
Unrelated TeamCity failure:
```TestPackageRenamesgithub.com/go-delve/delve/service/test```

**Comment by polinasok** [comment]
@derekparker Is there more to be done to get this PR in? It would be great to get it out of the way to resolve more conflicts in subsequent PRs. Thank you.

**Pull request merged**
